gitpod-io/gitpod
# [code/supervisor] allow retry auto exposing port

#### What it does

- fix #3160: display a user that ports failed to auto expose and allow to retry

Change in Gitpod Code: https://github.com/gitpod-io/vscode/commit/13852f7794240b9b6a25dfb321ac548c186212da

This PR also contains a commit which allows only one control port request in 2mins. I will drop it before merging. It is required for testing.

#### How to test

- Start a workspace and create 2 terminals.
- In one terminal run `curl lama.sh | LAMA_PORT=9000 sh`
- In another terminal run `curl lama.sh | LAMA_PORT=9001 sh`
- You should see that only one port will be detected and another will fail.
- In 2 mins after that, you should be able to click retry to expose on failed port and it will succeed as well.
